### ISSUE TITLE
BUG: Exclude From Search

### ISSUE THREAD

**Issue by gwhitehead** [10up/ElasticPress#3235]
### Describe the bug

I do like that there is the new Exclude from Search checkbox on each of the of the post editor screens. But it doesn't seem to issue a command to the elasticpress db for deleting the item from its system. A clearing of the index and rebuilding of the index doesn't seem to remove it from the autosuggest results nor the search results page. I have written a customization to the pre_get_posts so I can add the ID to an input field in the backend and have it not show up in the search results page but there doesn't appear to be a way to remove it from the autosuggest feature. I would like to think the new Exclude Search checkbox would remove it from both.

### Steps to Reproduce

1. Go into a post
2. mark it as exclude from search
3. rebuild index if so choose but would like to think the checkbox and save would fire a removal from ElasticPress DB and the search query

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gwhitehead** [comment]
So in my testing, even with this checkbox checked off it indexes the excluded page when rebuilding the index.  The only way to get an ID out of the autosuggest and search results is by utilizing the pre_get_posts exclusion based off of ID and then the filter attached to ep_index_posts_args on the building of IDs for the autosuggest. I would like to think when this exclude from search checkbox is checked off it not only removes it from the wp_query results but it also pings the elasticpress db and it removes the post ID from its database. I would also like to see a CLI way to request a removal of a post ID from the elasticpress db.

**Comment by felipeelia** [comment]
Hi @gwhitehead, thanks for opening the issue. The post is not deleted from Elasticsearch on purpose: we have it that way so people using the Protected Content feature can still find it while searching in the WP Dashboard. That is something we can change, just keeping it indexed if Protected Content is disabled, for example (a PR is welcomed.)

The feature is based on a meta field called `ep_exclude_from_search`. Do you have any code restricting which meta fields should be indexed? There is a chance that that field is not indexed, so EP can not use that to properly exclude the post from its search results.

**Comment by gwhitehead** [comment]
There are some limitations as to the meta fields but only ones  that end in a certain string, and _search or search is not one of them.  There is a ep_index_post_args filter that is removing post IDs at the indexing level.  There is a pre_get_posts action that is removing said IDs from the search results page, and then there is an ep_prepare_meta_data filter that is looking for meta_keys that end in a certain string like _alignment or _animation or _color just to name a few.

**Comment by felipeelia** [comment]
Okay @gwhitehead, do you mind sharing that custom code via https://pastebin.com/ or https://gist.github.com/ so we can give it a look? Also, it would be great if you could install [Debug Bar ElasticPress](https://10up.github.io/ElasticPress/tutorial-debugging.html), search for a term where an excluded post is returned and share all EP Debug Bar info (query args, results, etc.) via a pastebin here.

**Comment by gwhitehead** [comment]
Here are the filters/actions that we are running in the functions.php file. https://pastebin.com/CPYJVqM7

I am doing some javascript to customize the autosuggest but the autosuggest isn't the issue.

**Comment by gwhitehead** [comment]
Well things showing when excluded are an issue. But not the customization to the output.

**Comment by gwhitehead** [comment]
Just wanted to check in to see if there has been any resolution to the exclude from search checkbox actually removing items from results? It is mainly the removal from autosuggest that is needed without having to rebuild the database, which we have to do via the CLI in chunks due to its size.

**Comment by burhandodhy** [comment]
Hi @gwhitehead,

I'm unable to reproduce the issue. If the checkbox is selected, Elasticsearch excludes the post from the autosuggest list. I believe that some code is overriding the WordPress query. Please test this case without any plugins, and if the problem persists, please submit a bug report. I opened the enhancement ticket #3256  that will decrease the chances of query overriding.

**Comment by greg-carter1924** [comment]
I am having the same issue as @gwhitehead. I've checked the box on several products, yet the products appear in ElasticPress Autosuggest and Search Results. I have a separate ticket open with @felipeelia.

**Comment by gwhitehead** [comment]
> I am having the same issue as @gwhitehead. I've checked the box on several products, yet the products appear in ElasticPress Autosuggest and Search Results. I have a separate ticket open with @felipeelia.

@greg-carter1924,
I was able to track down my search results issue within my ep_prepare_meta_data filter.  I was excluding all meta data that started with an underscore and also found I had to check to see if my "not allowed" was not empty before unsetting a meta key. Search around your functions file to see if you are doing anything with the prepare_meta within your theme.
